### PR TITLE
Add Teams / Teams User management API

### PR DESCRIPTION
**Pull request by filipecabaco**
* Add token field to Team
* Add List Teams endpoint
* Add Show Team endpoint
* Add Create team endpoint
* Add Update team endpoint
* Add Delete endpoint
* Fix assertions on other endpoint tests

When listing or showing teams, we will consider the team the user is owner of and the ones he is part of via Team User.

Create, Update and Delete can only be done to the team they are owner of.

### CODE REVIEW

File: lib/logflare/teams.ex [hunk 1]
```diff
@@ -64,4 +68,40 @@ defmodule Logflare.Teams do
...
+      from t in Team,
+        join: tu in TeamUser,
+        on: tu.email == ^email
+
+    team_query =
+      from t in Team,
+        where: t.user_id == ^user_id,
+        union: ^team_user_query,
+        preload: [:user, :team_users]
+
```
Comment (Ziinc) [inline lib/logflare/teams.ex:85]: Can consolidate the query into:
```
from t in Team,
left_join: tu in assoc(t, :team_users)
where: tu.email == ^email or t.user_id == ^user_id
```
will be more performant than the union.

Also, should not preload by default, only conditionally. Each preload is an additional db query.

Also, using email might not be a good idea, not sure if the TeamUser `:token` is the same as the User `:token` when copied over, would be better to filter on it if possible.

File: lib/logflare_web/controllers/api/team_controller.ex [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  alias Logflare.Teams
+  action_fallback LogflareWeb.Api.FallbackController
+
+  def index(%{assigns: %{user: user}} = conn, _) do
+    teams = Teams.list_for_user(user)
+    json(conn, teams)
+  end
+
+  def show(%{assigns: %{user: user}} = conn, %{"token" => token}) do
+    with team when not is_nil(team) <- Teams.get(user, token),
```
Comment (Ziinc) [inline lib/logflare_web/controllers/api/team_controller.ex:12]: should handle situation when it is nil

File: lib/logflare_web/controllers/api/team_controller.ex [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    with team when not is_nil(team) <- Teams.get_team_by(token: token, user_id: user.id),
+         {:ok, _} <- Teams.delete_team(team) do
+      conn
+      |> Plug.Conn.send_resp(204, [])
+      |> Plug.Conn.halt()
+    end
+  end
+
+  defp params_keys_string_to_atom(params) do
+    params |> Enum.map(fn {k, v} -> {String.to_existing_atom(k), v} end) |> Enum.into(%{})
```
Comment (Ziinc) [inline lib/logflare_web/controllers/api/team_controller.ex:49]: why is this needed? Not very safe. ecto changeset should be used for casting.

File: lib/logflare_web/controllers/api/team_controller.ex [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+         {:ok, team} <- Teams.create_team(user, params),
+         team <- Teams.preload_fields(team, [:user, :team_users]) do
+      conn
+      |> put_status(201)
+      |> json(team)
+    end
+  end
+
+  def update(%{assigns: %{user: user}} = conn, %{"token" => token} = params) do
+    with params <- params_keys_string_to_atom(params),
```
Comment (Ziinc) [inline lib/logflare_web/controllers/api/team_controller.ex:29]: see comment below.

File: lib/logflare_web/controllers/api/team_controller.ex [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+  def show(%{assigns: %{user: user}} = conn, %{"token" => token}) do
+    with team when not is_nil(team) <- Teams.get(user, token),
+         team <- Teams.preload_fields(team, [:user, :team_users]) do
+      json(conn, team)
+    end
+  end
+
+  def create(%{assigns: %{user: user}} = conn, params) do
+    with params <- params_keys_string_to_atom(params),
```
Comment (Ziinc) [inline lib/logflare_web/controllers/api/team_controller.ex:19]: see comment below.

File: lib/logflare_web/router.ex [hunk 6]
```diff
@@ -359,6 +359,10 @@ defmodule LogflareWeb.Router do
     resources "/endpoints", Api.EndpointController,
       param: "token",
       only: [:index, :show, :create, :update, :delete]
+
+    resources "/teams", Api.TeamController,
+      param: "token",
+      only: [:index, :show, :create, :update, :delete]
```
Comment (Ziinc) [inline lib/logflare_web/router.ex:365]: can use `except: [:new, :edit]` instead

File: priv/repo/migrations/20221210011115_set_token_default_values.exs [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+
+  @disable_ddl_transaction true
+  @disable_migration_lock true
+
+  def change do
+    alter table(:teams) do
+      modify :token, :string, default: fragment("gen_random_uuid()")
+    end
+
+    execute "UPDATE teams SET token=gen_random_uuid()"
```
Comment (Ziinc) [inline priv/repo/migrations/20221210011115_set_token_default_values.exs:12]: this will replace all existing uuids already present. should filter to null.

File: test/logflare/teams_test.exs [hunk 8]
```diff
@@ -1,44 +1,91 @@
...
+    user = insert(:user)
+    teams = insert_list(2, :team_user, email: user.email)
+
+    expected = Enum.map(teams, & &1.team.id) |> Enum.sort()
+    result = Enum.map(Teams.list_for_user(user), & &1.id) |> Enum.sort()
+
+    assert expected == result
+  end
+
+  test "get/2 gets a team for a given user and token" do
```
Comment (Ziinc) [inline test/logflare/teams_test.exs:83]: should combine this function with `Teams.get_team/1` and pattern match on 2 arity

File: test/logflare_web/controllers/api/endpoint_controller_test.exs [hunk 9]
```diff
@@ -112,15 +112,15 @@ defmodule LogflareWeb.Api.EndpointControllerTest do
...
-      |> login_user(user)
-      |> delete("/api/endpoints/#{endpoint.token}", %{name: name})
-      |> response(204)
-
-      conn
-      |> login_user(user)
-      |> get("/api/endpoints/#{endpoint.token}")
-      |> response(404)
+      assert conn
+             |> login_user(user)
```
Comment (Ziinc) [inline test/logflare_web/controllers/api/endpoint_controller_test.exs:116]: can move `|> login_user(user)` to setup block

Comment (Ziinc) [reply test/logflare_web/controllers/api/endpoint_controller_test.exs:116]: or can use a setup function for :valid_user/:invalid_user logins

File: test/logflare_web/controllers/api/team_controller_test.exs [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+      non_owner_team: non_owner_team
+    } do
+      response =
+        conn
+        |> login_user(user)
+        |> get("/api/teams")
+        |> json_response(200)
+
+      response = response |> Enum.map(& &1["name"]) |> Enum.sort()
+      expected = [main_team, non_owner_team] |> Enum.map(& &1.name) |> Enum.sort()
```
Comment (Ziinc) [inline test/logflare_web/controllers/api/team_controller_test.exs:37]: we can do this instead to make it more concise.
```
assert [_, _] = json_resonse(conn, 200)
```
don't need to check equality, since we know there will only be 2 teams.

File: test/logflare_web/controllers/api/team_controller_test.exs [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+      user: user,
+      main_team: main_team
+    } do
+      response =
+        conn
+        |> login_user(user)
+        |> get("/api/teams/#{main_team.token}")
+        |> json_response(200)
+
+      assert response["name"] == main_team.name
```
Comment (Ziinc) [inline test/logflare_web/controllers/api/team_controller_test.exs:54]: can do instead to be more concise.
```
test " .... " %{..., main_team: %{id: team_id}}
...
assert %{"id"=> ^team_id} = json_respnse(conn, 200)
```

File: test/logflare_web/controllers/api/team_controller_test.exs [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+      invalid_user = insert(:user)
+
+      assert conn
+             |> login_user(invalid_user)
+             |> delete("/api/teams/#{main_team.token}")
+             |> response(404)
+    end
+  end
+
+  test "changeset errors handled gracefully", %{conn: conn, user: user, main_team: main_team} do
```
Comment (Ziinc) [inline test/logflare_web/controllers/api/team_controller_test.exs:154]: can be moved to the post/patch test cases since this is expected behaviour, will be less verbose.

**Review by Ziinc — CHANGES_REQUESTED** [review]
main issue is with the `String.to_existing_atom` usage, which looks unnecessary.

File: lib/logflare/teams.ex [hunk 13]
```diff
@@ -64,4 +68,40 @@ defmodule Logflare.Teams do
   @doc "Returns an `%Ecto.Changeset{}` for tracking team changes"
   @spec change_team(Team.t()) :: Ecto.Changeset.t()
   def change_team(%Team{} = team), do: Team.changeset(team, %{})
+
+  @doc "List all teams for user"
+  @spec list_for_user(User.t()) :: list(Team.t())
```
[the remark this answers is not in the document]
Comment (Ziinc) [reply lib/logflare/teams.ex:73]: could we also add in example usage for docs as well for the modified functions, so that we can incrementally improve the overall internal docs

File: lib/logflare/teams.ex (hunk 1 above)
Comment (filipecabaco) [reply lib/logflare/teams.ex:85]: won't work because `user_id` will be used to join `team_user`. Since the user isn't the owner of all `team_user` entities it won't find all the teams this user is a member of

`team_user` and `user` do not have matching tokens, we would need to add a new field to `team_user` with the `user` token to match with it which would require way more code changes

File: lib/logflare/teams.ex (hunk 13 above)
[the remark this answers is not in the document]
Comment (filipecabaco) [reply lib/logflare/teams.ex:73]: will add. 

[Team.t()] is the same as list(Team.t()) and more descriptive IMO

File: lib/logflare_web/controllers/api/team_controller.ex (hunk 2 above)
Comment (filipecabaco) [reply lib/logflare_web/controllers/api/team_controller.ex:12]: it is handled, it bubbles up to the action_fallback and returns 404

File: lib/logflare_web/controllers/api/team_controller.ex (hunk 3 above)
Comment (filipecabaco) [reply lib/logflare_web/controllers/api/team_controller.ex:49]: wasn't working for some reason 😞 unsure why

File: test/logflare/teams_test.exs (hunk 8 above)
Comment (filipecabaco) [reply test/logflare/teams_test.exs:83]: at the moment they have different purposes. `get_team/1` fetches the Team entity, `get/2` fetches the Team or a Team of a Team User the given user might be part of (that was confusing to write...)

but in the end, it has different purposes ... it might need a different name but it's a tricky one to handle

File: test/logflare_web/controllers/api/endpoint_controller_test.exs (hunk 9 above)
Comment (filipecabaco) [reply test/logflare_web/controllers/api/endpoint_controller_test.exs:116]: honestly don't seem much benefit, this way the tests are more descriptive of the flow they are executing for a given test

File: test/logflare_web/controllers/api/team_controller_test.exs (hunk 10 above)
Comment (filipecabaco) [reply test/logflare_web/controllers/api/team_controller_test.exs:37]: there are more teams created during setup ( the `_non_relevant_to_main_user`) to ensure we're fetching with authorization in mind hence the check of equality

File: test/logflare_web/controllers/api/team_controller_test.exs (hunk 11 above)
Comment (filipecabaco) [reply test/logflare_web/controllers/api/team_controller_test.exs:54]: can change 👍

File: lib/logflare/teams.ex (hunk 1 above)
Comment (filipecabaco) [reply lib/logflare/teams.ex:85]: actually I will try to left join with email vs union 👁️

File: lib/logflare_web/controllers/api/team_controller.ex (hunk 3 above)
Comment (filipecabaco) [reply lib/logflare_web/controllers/api/team_controller.ex:49]: stupid me it was something else for sure because it works... will fix the code

File: lib/logflare/teams.ex (hunk 13 above)
[the remark this answers is not in the document]
Comment (Ziinc) [reply lib/logflare/teams.ex:80]: 👍 both are fine to me

File: test/logflare_web/controllers/api/team_controller_test.exs (hunk 10 above)
Comment (Ziinc) [reply test/logflare_web/controllers/api/team_controller_test.exs:37]: i see. fetching a two len array would imply that it is working correctly as well.

if an equality check is really needed, would be simpler to check if name is in the response name array, as this is rather hard to read.

Comment (filipecabaco) [reply test/logflare_web/controllers/api/team_controller_test.exs:37]: oh good call! I can even get by token because I've added the token to the jason derive

File: lib/logflare/teams.ex (hunk 13 above)
[the remark this answers is not in the document]
Comment (filipecabaco) [reply lib/logflare/teams.ex:80]: I'm going to change, from the code I've seen we use more the version you mentioned

File: test/logflare/teams_test.exs (hunk 8 above)
Comment (Ziinc) [reply test/logflare/teams_test.exs:83]: seems like you're trying to do too much in one function, and also the function name doesn't self document and convey what is being fetched or checked for.

would suggest having a `Teams.has_team_access?/1` which accepts a User/TeamUser/email and does the check before actually fetching the team with `Teams.get_team/1`

File: lib/logflare/teams.ex (hunk 1 above)
Comment (Ziinc) [reply lib/logflare/teams.ex:92]: ok on 2nd point.

for 1st point, can join on the user email instead of using the assoc macro.

File: test/logflare/teams_test.exs (hunk 8 above)
Comment (filipecabaco) [reply test/logflare/teams_test.exs:83]: that would mean extra queries right? it's a weird balance because I want to fetch all the teams that user has in their dashboard which is the one they own and the ones they have been invited to... honestly I feel that the other methods are the ones wrongly named since they should be "owner" focused fetches 😅

File: lib/logflare/teams.ex (hunk 1 above)
Comment (filipecabaco) [reply lib/logflare/teams.ex:91]: joining on user email won't work either because the join will then apply the where clauses and fail:
```elixir
team_query =
      from t in Team,
        join: tu in TeamUser,
        on: tu.email == ^email,
        where: t.user_id == ^user_id,
        preload: [:user, :team_users]
```

File: lib/logflare/teams.ex (hunk 13 above)
[the remark this answers is not in the document]
Comment (Ziinc) [reply lib/logflare/teams.ex]: @filipecabaco could we update the name as well to be more in line with codegen conventions

File: test/logflare/teams_test.exs (hunk 8 above)
Comment (Ziinc) [reply test/logflare/teams_test.exs:83]: if thats the case then we can reuse the `list_teams_by_user_access` query and add a where filter to a team it to to one input team, that way we reuse the access logic. 

I think that having `Team.get_team/2` where first arg is the User/TeamUser is quite self explanatory that it is checking for whether user has access, though it is also ok to separate it out to two functions `Teams.get_team/1` which just retrieves any team, and `Teams.get_team_by_user_access/2` which then extends the has-access query. 

In general we opt for smaller, faster, and more flexible functions and less code dupe, because we can always use cachex if necessary for such rarely-updated tables.

Comment (Ziinc) [reply test/logflare/teams_test.exs:83]: in general we want to keep context functions as flexible as possible and don't want to put duplicate business logic in too many functions, as it makes refactoring or logic adjustments more complicated.

File: test/logflare_web/controllers/api/endpoint_controller_test.exs (hunk 9 above)
Comment (Ziinc) [reply test/logflare_web/controllers/api/endpoint_controller_test.exs:150]: sure, just seems like it is being repeated for each test

File: lib/logflare/teams.ex (hunk 1 above)
Comment (filipecabaco) [reply lib/logflare/teams.ex:91]: Screenshot 2023-02-02 at 11 26 25
still need to do a distinct and do the check on the email but this is closer 😄

File: test/logflare/teams_test.exs (hunk 8 above)
Comment (filipecabaco) [reply test/logflare/teams_test.exs:83]: I will extract the query from `list_teams_by_user_access` and then use composition to add the extra where clause on `get_team_by_user_access` 👀

File: lib/logflare/teams.ex (hunk 13 above)
[the remark this answers is not in the document]
Comment (filipecabaco) [reply lib/logflare/teams.ex]: changed the names to have that concept of `*_by_user_access`

File: priv/repo/migrations/20221210011115_set_token_default_values.exs (hunk 7 above)
Comment (Ziinc) [reply priv/repo/migrations/20221210011115_set_token_default_values.exs:12]: @filipecabaco i think should err on the side of not mutating existing data

**Review by Ziinc — APPROVED** [review]
changes look good, could you add in the migration where clause before merge? 💯 nice work

**Comment by filipecabaco** [comment]
Changed migrations and added a unique index to the token with the corresponding constraint
